packages/notifier/src/subscribe.js

```diff
@@ -26,25 +81,48 @@ export const subscribe = itP =>
  * longer needed so they can be garbage collected.
  *
  * @template T
- * @param {ERef<PublicationRecord<T>>} pubList
+ * @param {ERef<EachTopic<T>>} topic
+ * @param {ERef<PublicationRecord<T>>} nextCellP
+ *   PublicationRecord corresponding with the first iteration result
  * @returns {ForkableAsyncIterator<T, T>}
  */
-const makeEachIterator = pubList => {
+const makeEachIterator = (topic, nextCellP) => {
   // To understand the implementation, start with
   // https://web.archive.org/web/20160404122250/http://wiki.ecmascript.org/doku.php?id=strawman:concurrency#infinite_queue
   return Far('EachIterator', {
     next: () => {
-      const resultP = E.get(pubList).head;
-      pubList = E.get(pubList).tail;
+      const {
+        head: resultP,
+        publishCount: publishCountP,
+        tail: tailP,
+      } = E.get(nextCellP);
+
+      // If tailP is broken by upgrade, we will need to re-request it
+      // directly from `topic`.
+      const getSuccessor = async () => {
+        const publishCount = await publishCountP;
+        assert.typeof(publishCount, 'bigint');
+        const successor = await E(topic).subscribeAfter(publishCount);
+        const newPublishCount = successor.publishCount;
+        if (newPublishCount !== publishCount + 1n) {
+          Fail`eachIterator broken by gap from publishCount ${publishCount} to ${newPublishCount}`;
+        }
+        return successor;
+      };
+
+      // Replace nextCellP on every call to next() so things work even
+      // with an eager consumer that doesn't wait for results to settle.
+      nextCellP = reconnectAsNeeded(getSuccessor, [tailP]);
+
       // We expect the tail to be the "cannot read past end" error at the end
       // of the happy path.
       // Since we are wrapping that error with eventual send, we sink the
-      // rejection here too so it doesn't become an invalid unhandled rejection
-      // later.
-      void E.when(pubList, sink, sink);
+      // rejections here too to avoid invalid unhandled rejection issues later.
+      void E.when(publishCountP, sink, sink);
+      void E.when(nextCellP, sink, sink);
```

Comment:
I'm uncertain about why `publishCountP` is sunk, but `tailP` (originally `pubList` at this point) is not.  Can you elaborate (either by replying here, or in a code comment) as to why these two calls are necessary?

I suspect that there may be a clearer way to avoid introducing any unhandled rejections with minimal sinks.  I just don't see it right now.

Reply:
Any rejection of `tailP` would be handled in `reconnectAsNeeded`.